### ISSUE TITLE
ARC issues with cqlrt_cf

### ISSUE THREAD

**Issue by jackpal** [facebookincubator/CG-SQL#105]
I am running into problems trying to adopt the cqlrt_cf example runtime for use with the Swift Package Manager.

The Swift Package Manager lets you include C and Objective-C code in the package, but the Objective-C code is always compiled using ARC.

I can demonstrate the issues I'm seeing by adding the -fobjc-arc flag to the cqlrt_cf example:

```
cc -o demo -g -I.. -I. demo_todo.c demo_main.m cqlrt_cf.c cqlholder.m -lsqlite3 -framework Foundation -fobjc-arc

cqlholder.m:35:10: error: ARC forbids explicit message send of 'dealloc'
  [super dealloc];
   ~~~~~ ^
1 error generated.

```

If I update cqlholder.m:35
```obj-c
#if !__has_feature(objc_arc)
  [super dealloc];
#endif
```

Then the demo builds, but I get a segmentation fault when running the demo.

The segmentation fault is related to running this code:

```obj-c
  rs = CGS_todo_tasks_from_todo_tasks(result_set);
```

Any ideas how to proceed?

**Comment by ricomariani** [comment]
Yeah, you don't use dealloc at all in ARC.  I'm not sure what the fault is.  I'm on vacation next week but I'll see what I can do when I'm back.  The other thing you asked about is pretty easy to do.

**Comment by ricomariani** [comment]
Hello world looks like this with ARC.

```
#import <Foundation/Foundation.h>

int main (int argc, const char * argv[])
{
        NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];

        NSLog (@"hello world");
        [pool drain];
        return 0;
}
```

**Comment by ricomariani** [comment]
Maybe change demo_main accordingly.

**Comment by ricomariani** [comment]
We probably never need to call dealloc anyway since we know we're just a subtype of NSObject anyway.

**Comment by jackpal** [comment]
I created an Xcode project for demo, and ran with address sanitizer and zombie detection enabled, and I think I found the issue:

In cqlholder.m:86
```obj-c
  return (__bridge cql_result_set_ref)holder;
```

Under ARC, this line deallocates holder while returning it.
If I change the code to:

```obj-c
  return (__bridge_retained cql_result_set_ref)holder;
```

Then everything works.

I wonder if `cql_box_stmt` needs a similar change. (I haven't observed an issue with cql_box_stmt, though.)

**Comment by jackpal** [comment]
FWIW "NSAutoReleasePool" doesn't compile with ARC, you have to use the "@autoreleasepool { }" version. But that by itself doesn't fix the issue.

**Comment by ricomariani** [comment]
First some diagnostics:

```
 - (instancetype)initWithBytes:(void *)bytesIn andType:(int)typeIn
 {
+  NSLog(@"init!");
   self = [super init];
   self.bytes = bytesIn;
   self.type = typeIn;
@@ -28,11 +31,15 @@
 
 - (void)dealloc
 {
+  NSLog(@"dealloc!");
   if (self.bytes)
   {
     [self dynamicTeardown];
   }
+
+#if !__has_feature(objc_arc)
   [super dealloc];
+#endif
 }
 ```
 
 The super dealloc thing you already knew.  I don' thinks needed in either version actually because the base is NSObject (i.e. nothing) so no dealloc should be needed.  But anyway.

Now the fix:

```
 // The two kinds of things we know how to hold.
@@ -83,7 +90,11 @@
   result_set->meta = meta;
 
   CQLHolder *holder = [[CQLHolder alloc] initWithBytes:(void *)result_set andType:CF_HELD_TYPE_RESULT_SET];
+#if !__has_feature(objc_arc)
   return (__bridge cql_result_set_ref)holder;
+#else
+  return (__bridge_retained cql_result_set_ref)holder;
+#endif
 }
 
 cql_object_ref _Nonnull cql_box_stmt(sqlite3_stmt *_Nullable stmt)
@@ -92,7 +103,11 @@
   box->stmt = stmt;
 
   CQLHolder *holder = [[CQLHolder alloc] initWithBytes:(void *)box andType:CF_HELD_TYPE_BOXED_STMT];
+#if !__has_feature(objc_arc)
   return (__bridge cql_object_ref)holder;
+#else
+  return (__bridge_retained cql_object_ref)holder;
+#endif
 }
```
 
With ARC the create functions must use `__bridge_retained`

**Comment by ricomariani** [comment]
To get correct cleanup you must also use

```
 @autoreleasepool {

  // the body of main
 
 }
 
 ```

**Comment by ricomariani** [comment]
Since non-ARC is rarely used I'll change the demo so that it opts into ARC.  I could leave the ifdefs but I think it will just confuse people.

**Comment by jackpal** [comment]
SGTM!

**Comment by ricomariani** [comment]
```
diff --git a/xplat/msys/tools/cql/cqlrt_cf/cqlholder.m b/xplat/msys/tools/cql/cqlrt_cf/cqlholder.m
--- a/xplat/msys/tools/cql/cqlrt_cf/cqlholder.m
+++ b/xplat/msys/tools/cql/cqlrt_cf/cqlholder.m
@@ -32,7 +32,6 @@
   {
     [self dynamicTeardown];
   }
-  [super dealloc];
 }
 
 // The two kinds of things we know how to hold.
@@ -56,7 +55,7 @@
       cql_boxed_stmt *box = (cql_boxed_stmt *)self.bytes;
 
       // note this is a no-op on null statements, and nulls stmt after finalization
-      cql_finalize_stmt(&box->stmt);  
+      cql_finalize_stmt(&box->stmt);
       break;
     }
   }
@@ -83,7 +82,7 @@
   result_set->meta = meta;
 
   CQLHolder *holder = [[CQLHolder alloc] initWithBytes:(void *)result_set andType:CF_HELD_TYPE_RESULT_SET];
-  return (__bridge cql_result_set_ref)holder;
+  return (__bridge_retained cql_result_set_ref)holder;
 }
 
 cql_object_ref _Nonnull cql_box_stmt(sqlite3_stmt *_Nullable stmt)
@@ -92,7 +91,7 @@
   box->stmt = stmt;
 
   CQLHolder *holder = [[CQLHolder alloc] initWithBytes:(void *)box andType:CF_HELD_TYPE_BOXED_STMT];
-  return (__bridge cql_object_ref)holder;
+  return (__bridge_retained cql_object_ref)holder;
 }
 
 // For holding boxed statements.
@@ -105,7 +104,7 @@
 }
 
 cql_result_set *_Nonnull cql_get_result_set_from_ref(cql_result_set_ref _Nonnull ref)
-{ 
+{
   CQLHolder *holder = (__bridge CQLHolder *)ref;
   cql_result_set *_Nonnull result_set = (cql_result_set *)holder.bytes;
   return result_set;
diff --git a/xplat/msys/tools/cql/cqlrt_cf/demo_main.m b/xplat/msys/tools/cql/cqlrt_cf/demo_main.m
--- a/xplat/msys/tools/cql/cqlrt_cf/demo_main.m
+++ b/xplat/msys/tools/cql/cqlrt_cf/demo_main.m
@@ -12,84 +12,78 @@
 
 int main(int argc, char **argv)
 {
-  /* Note: not exactly world class error handling but that isn't the point */
 
+  @autoreleasepool {
 
-  // create a db
-  sqlite3 *db;
-  int rc = sqlite3_open(":memory:", &db);
-  if (rc != SQLITE_OK) {
-    exit(1);
-  }
+    // Note: not exactly world class error handling but that isn't the point
 
-  // make schema if needed (it always will be here because memory databases begin empty
-  rc = todo_create_tables(db);
-   if (rc != SQLITE_OK) {
-    exit(2);
-  }
+    // create a db
+    sqlite3 *db;
+    int rc = sqlite3_open(":memory:", &db);
+    if (rc != SQLITE_OK) {
+      exit(1);
+    }
+
+    // make schema if needed (it always will be here because memory databases begin empty
+    rc = todo_create_tables(db);
+     if (rc != SQLITE_OK) {
+      exit(2);
+    }
 
-  // add some tasks
-  const char *const default_tasks[] = {
-    "Buy milk",
-    "Walk dog",
-    "Write code"
-  };
-
-  for (int i = 0; i < 3; i++) {
-    // note we make a string reference from a c string here
-    cql_string_ref dtask = cql_string_ref_new(default_tasks[i]);
-    rc = todo_add(db, dtask);
-    cql_string_release(dtask); // and then dispose of the reference
+    // add some tasks
+    const char *const default_tasks[] = {
+      "Buy milk",
+      "Walk dog",
+      "Write code"
+    };
+
+    for (int i = 0; i < 3; i++) {
+      // note we make a string reference from a c string here
+      cql_string_ref dtask = cql_string_ref_new(default_tasks[i]);
+      rc = todo_add(db, dtask);
+      cql_string_release(dtask); // and then dispose of the reference
+      if (rc != SQLITE_OK) {
+        exit(3);
+      }
+    }
+
+    // mark a task as done
+    rc = todo_setdone_(db, 1, true);
     if (rc != SQLITE_OK) {
-      exit(3);
+      exit(4);
     }
-  }
 
-  // mark a task as done
-  rc = todo_setdone_(db, 1, true);
-  if (rc != SQLITE_OK) {
-    exit(4);
-  }
+    // delete a row in the middle, rowid = 2
+    rc = todo_delete(db, 2);
+    if (rc != SQLITE_OK) {
+      exit(5);
+    }
 
-  // delete a row in the middle, rowid = 2
-  rc = todo_delete(db, 2);
-  if (rc != SQLITE_OK) {
-    exit(5);
-  }
+    // select out some results
+    todo_tasks_result_set_ref result_set_ref;
+    rc = todo_tasks_fetch_results(db, &result_set_ref);
+    if (rc != SQLITE_OK) {
+      printf("error: %d\n", rc);
+      exit(2);
+    }
 
-  CGS_todo_tasks *rs = NULL;
+    CGS_todo_tasks *rs = CGS_todo_tasks_from_todo_tasks(result_set_ref);
+    cql_release(result_set_ref);  // objective C is now in control of the refs
 
-  // select out some results
-  todo_tasks_result_set_ref result_set;
-  rc = todo_tasks_fetch_results(db, &result_set);
-  if (rc != SQLITE_OK) {
-    printf("error: %d\n", rc);
-    exit(2);
-  }
-  
-  // Note that this is not a create/copy method, objc knows
-  // so rs will not be released. You have to use ARC tricks
-  // if you want a retain here but usually you don't.  You
-  // can simply pass rs around at this point and it will retain
-  // if it needs to in the usual way.
-  rs = CGS_todo_tasks_from_todo_tasks(result_set);
-  
-  // get result count
-  cql_int32 result_count = CGS_todo_tasks_result_count(rs);
-  
-  // loop to print
-  for (cql_int32 row = 0; row < result_count; row++) {
-    // note the usual "get" semantics, no retain here
-    NSString *text = CGS_todo_tasks_get_description(rs, row);
-    cql_bool done = CGS_todo_tasks_get_done(rs, row);
-    cql_int32 rowid = CGS_todo_tasks_get_rowid(rs, row);
-  
-    NSLog(@"%d: rowid:%d %@ (%s)\n", row, rowid, text, done ? "done" : "not done");
-  }
+    // get result count
+    cql_int32 result_count = CGS_todo_tasks_result_count(rs);
 
-  // this is going to cause the dealloc to run (rs = NULL won't do the job)
-  cql_release(result_set);
+    // loop to print
+    for (cql_int32 row = 0; row < result_count; row++) {
+      // note the usual "get" semantics, no retain here
+      NSString *text = CGS_todo_tasks_get_description(rs, row);
+      cql_bool done = CGS_todo_tasks_get_done(rs, row);
+      cql_int32 rowid = CGS_todo_tasks_get_rowid(rs, row);
 
-  // and close the database
-  sqlite3_close(db);
+      NSLog(@"%d: rowid:%d %@ (%s)\n", row, rowid, text, done ? "done" : "not done");
+    }
+
+    // and close the database
+    sqlite3_close(db);
+  }
 }
diff --git a/xplat/msys/tools/cql/cqlrt_cf/make.sh b/xplat/msys/tools/cql/cqlrt_cf/make.sh
--- a/xplat/msys/tools/cql/cqlrt_cf/make.sh
+++ b/xplat/msys/tools/cql/cqlrt_cf/make.sh
@@ -21,7 +21,7 @@
 ../out/cql --in demo_todo.sql --cg demo_objc.h --rt objc_mit --objc_c_include_path demo_todo.h --cqlrt cqlrt_cf.h
 
 echo "building executable"
-cc -o demo -g -I.. -I. demo_todo.c demo_main.m cqlrt_cf.c cqlholder.m -lsqlite3 -framework Foundation
+cc -o demo -g -I.. -I. demo_todo.c demo_main.m cqlrt_cf.c cqlholder.m -lsqlite3 -framework Foundation -fobjc-arc
 
 echo "running demo"
 ./demo
```

**Comment by ricomariani** [comment]
That's my patch (it's mostly indent change)

**Comment by jackpal** [comment]
One unrelated fix you can do as long as you're updating main.m (I noticed this warning in Xcode)

```obj-c
    cql_int32 rowid = CGS_todo_tasks_get_rowid(rs, row);
```

should be

```obj-c
    cql_int64 rowid = CGS_todo_tasks_get_rowid(rs, row);
```

**Comment by ricomariani** [comment]
Got it because int64.

**Comment by ricomariani** [comment]
All the above is just waiting on code review now.
________________________________

**Comment by ricomariani** [comment]
This has all landed.

**Comment by jackpal** [comment]
Thanks! I've updated to it, and it's working great.
